Portainer panics when starting off an existing EE database without data migration

<!--

Thanks for reporting a bug for Portainer !

You can find more information about Portainer support framework policy here: https://www.portainer.io/2019/04/portainer-support-policy/

Do you need help or have a question? Come chat with us on Slack http://portainer.io/slack/.

Before opening a new issue, make sure that we do not have any duplicates
already open. You can ensure this by searching the issue list for this
repository. If there is a duplicate, please close your issue and add a comment
to the existing issue instead.

Also, be sure to check our FAQ and documentation first: https://documentation.portainer.io/
-->

**Bug description**

Starting Portainer CE from an existing EE database without going through the database migration steps will create a Go panic.


**Expected behavior**

Portainer should display an error message if it is unable to go through the database migration process and provide a link to the database migration documentation.

**Portainer Logs**

```
2020/11/09 02:21:30 Migrating database from version 1 to 25.
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x8 pc=0x5e08ca]

goroutine 1 [running]:
github.com/boltdb/bolt.(*Bucket).Cursor(...)
	/tmp/go/pkg/mod/github.com/boltdb/bolt@v1.3.1/bucket.go:91
github.com/portainer/portainer/api/bolt/migrator.(*Migrator).retrieveLegacyResourceControls.func1(0xc000130ee0, 0x0, 0xc000130ee0)
	/home/vsts/work/1/s/api/bolt/migrator/migrate_dbversion1.go:66 +0xaa
github.com/boltdb/bolt.(*DB).View(0xc00000a960, 0xc0005359b8, 0x0, 0x0)
	/tmp/go/pkg/mod/github.com/boltdb/bolt@v1.3.1/db.go:629 +0xa8
github.com/portainer/portainer/api/bolt/migrator.(*Migrator).retrieveLegacyResourceControls(0xc000535b80, 0x3d, 0x40, 0x3d, 0x0, 0x0)
	/home/vsts/work/1/s/api/bolt/migrator/migrate_dbversion1.go:64 +0x82
github.com/portainer/portainer/api/bolt/migrator.(*Migrator).updateResourceControlsToDBVersion2(0xc000535b80, 0x0, 0x0)
	/home/vsts/work/1/s/api/bolt/migrator/migrate_dbversion1.go:10 +0x40
github.com/portainer/portainer/api/bolt/migrator.(*Migrator).Migrate(0xc000535b80, 0x2a, 0xc000535b60)
	/home/vsts/work/1/s/api/bolt/migrator/migrator.go:105 +0x7e6
github.com/portainer/portainer/api/bolt.(*Store).MigrateData(0xc0000b81c0, 0x0, 0x0)
	/home/vsts/work/1/s/api/bolt/datastore.go:157 +0x3d1
main.initDataStore(0x19dc1cc, 0x5, 0x1cb0100, 0xc0001d0f60, 0xc0000a2200, 0x5)
	/home/vsts/work/1/s/api/cmd/portainer/main.go:67 +0x156
main.main()
	/home/vsts/work/1/s/api/cmd/portainer/main.go:340 +0x99
```

**Steps to reproduce the issue:**

1. Deploy an instance of Portainer EE
2. Stop that instance
3. Deploy an instance of Portainer CE reusing the data from the EE instance

**Technical details:**

- Portainer version: 2.0-CE

--- a/api/bolt/datastore.go
+++ b/api/bolt/datastore.go
@@ -6,7 +6,7 @@ import (
 	"time"
 
 	"github.com/boltdb/bolt"
-	"github.com/portainer/portainer/api"
+	portainer "github.com/portainer/portainer/api"
 	"github.com/portainer/portainer/api/bolt/customtemplate"
 	"github.com/portainer/portainer/api/bolt/dockerhub"
 	"github.com/portainer/portainer/api/bolt/edgegroup"
@@ -69,6 +69,14 @@ type Store struct {
 	WebhookService          *webhook.Service
 }
 
+func (store *Store) edition() portainer.SoftwareEdition {
+	edition, err := store.VersionService.Edition()
+	if err == errors.ErrObjectNotFound {
+		edition = portainer.PortainerCE
+	}
+	return edition
+}
+
 // NewStore initializes a new Store and the associated services
 func NewStore(storePath string, fileService portainer.FileService) (*Store, error) {
 	store := &Store{
@@ -114,6 +122,14 @@ func (store *Store) Close() error {
 // existing data.
 func (store *Store) IsNew() bool {
 	return store.isNew
+}
+
+// CheckCurrentEdition checks if current edition is community edition
+func (store *Store) CheckCurrentEdition() error {
+	if store.edition() != portainer.PortainerCE {
+		return errors.ErrWrongDBEdition
+	}
+	return nil
 }
 
 // MigrateData automatically migrate the data based on the DBVersion.

--- a/api/bolt/errors/errors.go
+++ b/api/bolt/errors/errors.go
@@ -4,4 +4,5 @@ import "errors"
 
 var (
 	ErrObjectNotFound = errors.New("Object not found inside the database")
+	ErrWrongDBEdition = errors.New("The Portainer database is set for Portainer Business Edition, please follow the instructions in our documention to downgrade it: https://documentation.portainer.io/v2.0-be/downgrade/be-to-ce/")
 )

--- a/api/bolt/version/version.go
+++ b/api/bolt/version/version.go
@@ -4,6 +4,7 @@ import (
 	"strconv"
 
 	"github.com/boltdb/bolt"
+	portainer "github.com/portainer/portainer/api"
 	"github.com/portainer/portainer/api/bolt/errors"
 	"github.com/portainer/portainer/api/bolt/internal"
 )
@@ -13,6 +14,7 @@ const (
 	BucketName  = "version"
 	versionKey  = "DB_VERSION"
 	instanceKey = "INSTANCE_ID"
+	editionKey  = "EDITION"
 )
 
 // Service represents a service to manage stored versions.
@@ -56,6 +58,21 @@ func (service *Service) DBVersion() (int, error) {
 	return strconv.Atoi(string(data))
 }
 
+// Edition retrieves the stored portainer edition.
+func (service *Service) Edition() (portainer.SoftwareEdition, error) {
+	editionData, err := service.getKey(editionKey)
+	if err != nil {
+		return 0, err
+	}
+
+	edition, err := strconv.Atoi(string(editionData))
+	if err != nil {
+		return 0, err
+	}
+
+	return portainer.SoftwareEdition(edition), nil
+}
+
 // StoreDBVersion store the database version.
 func (service *Service) StoreDBVersion(version int) error {
 	return service.db.Update(func(tx *bolt.Tx) error {
@@ -97,5 +114,38 @@ func (service *Service) StoreInstanceID(ID string) error {
 
 		data := []byte(ID)
 		return bucket.Put([]byte(instanceKey), data)
+	})
+}
+
+func (service *Service) getKey(key string) ([]byte, error) {
+	var data []byte
+
+	err := service.db.View(func(tx *bolt.Tx) error {
+		bucket := tx.Bucket([]byte(BucketName))
+
+		value := bucket.Get([]byte(key))
+		if value == nil {
+			return errors.ErrObjectNotFound
+		}
+
+		data = make([]byte, len(value))
+		copy(data, value)
+
+		return nil
+	})
+
+	if err != nil {
+		return nil, err
+	}
+
+	return data, nil
+}
+
+func (service *Service) setKey(key string, value string) error {
+	return service.db.Update(func(tx *bolt.Tx) error {
+		bucket := tx.Bucket([]byte(BucketName))
+
+		data := []byte(value)
+		return bucket.Put([]byte(key), data)
 	})
 }

--- a/api/cmd/portainer/main.go
+++ b/api/cmd/portainer/main.go
@@ -375,6 +375,10 @@ func main() {
 	dataStore := initDataStore(*flags.Data, fileService)
 	defer dataStore.Close()
 
+	if err := dataStore.CheckCurrentEdition(); err != nil {
+		log.Fatal(err)
+	}
+
 	jwtService, err := initJWTService(dataStore)
 	if err != nil {
 		log.Fatal(err)

--- a/api/portainer.go
+++ b/api/portainer.go
@@ -559,6 +559,9 @@ type (
 	// SnapshotJob represents a scheduled job that can create endpoint snapshots
 	SnapshotJob struct{}
 
+	// SoftwareEdition represents an edition of Portainer
+	SoftwareEdition int
+
 	// Stack represents a Docker stack created via docker stack deploy
 	Stack struct {
 		ID              StackID          `json:"Id"`
@@ -822,6 +825,7 @@ type (
 		Close() error
 		IsNew() bool
 		MigrateData() error
+		CheckCurrentEdition() error
 
 		DockerHub() DockerHubService
 		CustomTemplate() CustomTemplateService
@@ -1122,8 +1126,9 @@ type (
 	// VersionService represents a service for managing version data
 	VersionService interface {
 		DBVersion() (int, error)
-		StoreDBVersion(version int) error
+		Edition() (SoftwareEdition, error)
 		InstanceID() (string, error)
+		StoreDBVersion(version int) error
 		StoreInstanceID(ID string) error
 	}
 
@@ -1266,6 +1271,16 @@ const (
 	TeamLeader
 	// TeamMember represents a member role inside a team
 	TeamMember
+)
+
+const (
+	_ SoftwareEdition = iota
+	// PortainerCE represents the community edition of Portainer
+	PortainerCE
+	// PortainerBE represents the business edition of Portainer
+	PortainerBE
+	// PortainerEE represents the business edition of Portainer
+	PortainerEE
 )
 
 const (

